### PR TITLE
feat: redact store values from logs

### LINKED ISSUE #1173: Obscure pepr store calls / add a "sensitive" parameter
### Is your feature request related to a problem? Please describe.

In uds-core we are using the Pepr Store to store Keycloak client registration tokens (sensitive). When we make a call to put the value in the store, the value is logged out if the log level is set to DEBUG or higher.

### Describe the solution you'd like

If a setItem call is made to the store, then Pepr does not log out the value, just the key.

### Describe alternatives you've considered

Alternatively, when the "sensitive" parameter is optionally passed to `setItem`, then the value is obfuscated (i.e. `***`) in logs.

### Additional context

In order to avoid this issue in uds core we reduced the log level of pepr, but in some cases it is nice to increase that log level temporarily for debugging. It would be ideal if that did not expose any sensitive values.

### PR DESCRIPTION
**Pull request by cmwylie19**
## Description

Optionally do not expose store values in logs

End to End Test:  https://github.com/defenseunicorns/pepr-excellent-examples/pull/78

```plaintext
[15:54:39.347] DEBUG (19402): Sending updates to Pepr store
    add:/data/hello-pepr-v2-password:**redacted**: {
      "op": "add",
      "path": "/data/hello-pepr-v2-password",
      "value": "**redacted**"
    }
[15:54:39.373] DEBUG (19402): Pepr Store update
    apiVersion: "pepr.dev/v1"
    data: {
      "__pepr_do_not_delete__": "**redacted**",
      "hello-pepr-v2-password": "**redacted**"
    }
    kind: "PeprStore"
    metadata: {
      "creationTimestamp": "2024-09-26T19:54:36Z",
      "generation": 2,
      "managedFields": [
        {
          "apiVersion": "pepr.dev/v1",
          "fieldsType": "FieldsV1",
          "fieldsV1": {
            "f:data": {
              "f:__pepr_do_not_delete__": {}
            }
          },
          "manager": "pepr",
          "operation": "Apply",
          "time": "2024-09-26T19:54:36Z"
        },
        {
          "apiVersion": "pepr.dev/v1",
          "fieldsType": "FieldsV1",
          "fieldsV1": {
            "f:data": {
              "f:hello-pepr-v2-password": {}
            }
          },
          "manager": "kubernetes-fluent-client",
          "operation": "Update",
          "time": "2024-09-26T19:54:39Z"
        }
      ],
      "name": "pepr-static-test-store",
      "namespace": "pepr-system",
      "resourceVersion": "1088",
      "uid": "b51c4a6a-2150-4e0a-8254-55a5687c443d"
    }
[15:54:44.375] INFO (19402): getItem
    password: "***DECLASSIFIED***"
[15:54:44.375] INFO (19402): DONE
[15:54:49.349] DEBUG (19402): Sending updates to Pepr store
    remove:**redacted**: {
      "op": "remove",
      "path": "/data/hello-pepr-v2-password"
    }
[15:54:49.366] DEBUG (19402): Pepr Store update
    apiVersion: "pepr.dev/v1"
    data: {
      "__pepr_do_not_delete__": "**redacted**"
    }
    kind: "PeprStore"
    metadata: {
      "creationTimestamp": "2024-09-26T19:54:36Z",
      "generation": 3,
      "managedFields": [
```

## Related Issue

Fixes #1173 

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging
- [x] Unit, [Journey](https://github.com/defenseunicorns/pepr/tree/main/journey), [E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples), [docs](https://github.com/defenseunicorns/pepr/tree/main/docs), [adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or updated as needed
- [x] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

### CODE REVIEW

File: docs/030_user-guide/120_customization.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
...
 This document outlines how to customize the build output through Helm overrides and `package.json` configurations.
 
+## Redact Store Values from Logs
+
+By default, the store values are displayed in logs, to redact them you can set the `PEPR_STORE_REDACTED_VALUES` environment variable to `true` in the `package.json` file or directly on the Watcher or Admission `Deployment`. The default value is `undefined`.
+
+```json
+{
+  "env": {
+    "PEPR_STORE_REDACTED_VALUES": "true"
```
Comment (cmwylie19) [inline docs/030_user-guide/120_customization.md]:
```suggestion
    "PEPR_STORE_REDACT_VALUES": "true"
```

File: docs/030_user-guide/120_customization.md [hunk 2]
```diff
@@ -2,6 +2,18 @@
 
 This document outlines how to customize the build output through Helm overrides and `package.json` configurations.
 
+## Redact Store Values from Logs
+
+By default, the store values are displayed in logs, to redact them you can set the `PEPR_STORE_REDACTED_VALUES` environment variable to `true` in the `package.json` file or directly on the Watcher or Admission `Deployment`. The default value is `undefined`.
```
Comment (cmwylie19) [inline docs/030_user-guide/120_customization.md]:
```suggestion
By default, the store values are displayed in logs, to redact them you can set the `PEPR_STORE_REDACT_VALUES` environment variable to `true` in the `package.json` file or directly on the Watcher or Admission `Deployment`. The default value is `undefined`.
```

File: src/lib/controller/store.test.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  {
+    op: "add",
+    path: "add:/data/hello-pepr-v2-f",
+    value: "secret",
+  },
+];
+
+describe("redactedPatch", () => {
+  afterEach(() => {
+    delete process.env.PEPR_STORE_REDACTED_VALUES;
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
    delete process.env.PEPR_STORE_REDACT_VALUES;
```

File: src/lib/controller/store.test.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  test("should redact keys and values if PEPR_STORE_REDACTED_VALUES is true", () => {
+    process.env.PEPR_STORE_REDACTED_VALUES = "true";
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
    process.env.PEPR_STORE_REDACT_VALUES = "true";
```

File: src/lib/controller/store.ts [hunk 5]
```diff
@@ -305,3 +306,36 @@ export class PeprControllerStore {
     }
   };
 }
+
+export function redactedStore(store: PeprStore): PeprStore {
+  const redacted = process.env.PEPR_STORE_REDACTED_VALUES === "true";
```
Comment (cmwylie19) [inline src/lib/controller/store.ts]:
```suggestion
  const redacted = process.env.PEPR_STORE_REDACT_VALUES === "true";
```

File: src/lib/controller/store.ts [hunk 6]
```diff
@@ -305,3 +306,36 @@ export class PeprControllerStore {
...
+    ...store,
+    data: Object.keys(store.data).reduce((acc: Record<string, string>, key: string) => {
+      acc[key] = redacted ? redactedValue : store.data[key];
+      return acc;
+    }, {}),
+  };
+}
+
+export function redactedPatch(patch: Record<string, Operation> = {}): Record<string, Operation> {
+  const redacted = process.env.PEPR_STORE_REDACTED_VALUES === "true";
```
Comment (cmwylie19) [inline src/lib/controller/store.ts]:
```suggestion
  const redacted = process.env.PEPR_STORE_REDACT_VALUES === "true";
```

File: src/lib/controller/store.test.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    namespace: fc.string(),
+  }),
+  data: fc.dictionary(
+    fc.string().filter(str => str !== "__proto__"),
+    fc.string().filter(str => str !== "__proto__"),
+  ),
+});
+describe("Fuzzing redactedStore", () => {
+  afterEach(() => {
+    delete process.env.PEPR_STORE_REDACTED_VALUES;
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
    delete process.env.PEPR_STORE_REDACT_VALUES;
```

File: src/lib/controller/store.test.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  test("should redact values if PEPR_STORE_REDACTED_VALUES is true", () => {
+    fc.assert(
+      fc.property(peprStoreFuzz, store => {
+        process.env.PEPR_STORE_REDACTED_VALUES = "true";
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
        process.env.PEPR_STORE_REDACT_VALUES = "true";
```

File: src/lib/controller/store.test.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    fc.string().filter(str => str !== "__proto__"),
+    fc.string().filter(str => str !== "__proto__"),
+  ),
+});
+describe("Fuzzing redactedStore", () => {
+  afterEach(() => {
+    delete process.env.PEPR_STORE_REDACTED_VALUES;
+  });
+
+  test("should redact values if PEPR_STORE_REDACTED_VALUES is true", () => {
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
  test("should redact values if PEPR_STORE_REDACT_VALUES is true", () => {
```

File: src/lib/controller/store.test.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        const result = redactedStore(store);
+
+        Object.values(result.data).forEach(value => {
+          expect(value).toBe(redactedValue);
+        });
+      }),
+    );
+  });
+
+  test("should not redact values if PEPR_STORE_REDACTED_VALUES is not true", () => {
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
  test("should not redact values if PEPR_STORE_REDACT_VALUES is not true", () => {
```

File: src/lib/controller/store.test.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    fc.assert(
+      fc.property(peprStoreFuzz, store => {
+        process.env.PEPR_STORE_REDACTED_VALUES = "false";
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
        process.env.PEPR_STORE_REDACT_VALUES = "false";
```

File: src/lib/controller/store.test.ts [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        expect(result.data).toEqual(store.data);
+      }),
+    );
+  });
+
+  test("should maintain other properties of the store", () => {
+    fc.assert(
+      fc.property(peprStoreFuzz, store => {
+        const redactionEnabled = fc.boolean();
+        process.env.PEPR_STORE_REDACTED_VALUES = redactionEnabled ? "true" : "false";
```
Comment (cmwylie19) [inline src/lib/controller/store.test.ts]:
```suggestion
        process.env.PEPR_STORE_REDACT_VALUES = redactionEnabled ? "true" : "false";
```

File: src/lib/controller/store.test.ts [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    value: "secret",
+  },
+];
+
+describe("redactedPatch", () => {
+  afterEach(() => {
+    delete process.env.PEPR_STORE_REDACT_VALUES;
+  });
+
+  test("should redact keys and values if PEPR_STORE_REDACTED_VALUES is true", () => {
```
Comment (btlghrants) [inline src/lib/controller/store.test.ts]:
```suggestion
  test("should redact keys and values if PEPR_STORE_REDACT_VALUES is true", () => {
```

File: src/lib/controller/store.test.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    process.env.PEPR_STORE_REDACT_VALUES = "true";
+    addOperationKeys.forEach((key, i) => {
+      const redactedResult = redactedPatch({ [key]: addOperationValues[i] });
+      for (const [k, v] of Object.entries(redactedResult)) {
+        expect(k).toContain(`:${redactedValue}`);
+        expect(v).toEqual(expect.objectContaining({ value: redactedValue }));
+      }
+    });
+  });
+  test("should not redact keys and values if PEPR_STORE_REDACTED_VALUES is not true", () => {
```
Comment (btlghrants) [inline src/lib/controller/store.test.ts]:
```suggestion
  test("should not redact keys and values if PEPR_STORE_REDACT_VALUES is not true", () => {
```

**Review by btlghrants — COMMENTED** [review]
After the demo / walkthrough this morning -- and (mostly) renaming the env var -- it looks good; should meet needs while still allowing us easy access to values in the e2es.

Just need to roll-in those two places in the test titles where the rename didn't happen & I'll approve.

**Review by btlghrants — APPROVED** [review]
n/m -- I was able to commit the suggestions myself after the review completed.  Test titles updated & LGTM.

**Pull request merged**